# Art Works Analysis
dataset: [art works](https://www.tate.org.uk/about-us/digital/collection-data)

In [ ]:
import pandas as pd

In [9]:
data = pd.read_csv("Data/artwork_sample.csv")
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine    year  acquisitionYear  \
0                Presented by Mrs John Richmond 1922     NaN             1922   
1                Presented by Mrs John Richmond 1922     NaN             1922   
2                Presented by Mrs John Richmond 1922  1785.0             1922   
3                Presented by Mrs John Richmond 1922     NaN             1922   
4  Purchased with the assistance of a special gra...  1826.0             1919   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   
2  support: 343 x 467 mm    343     467    NaN    mm          NaN   
3  support: 318 x 394 mm    318     394    NaN    mm          NaN   
4    image: 243 x 335 mm    243     335    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...  
2  http://www.tate.org.uk/art/artworks/blake-the-...  
3  http://www.tate.org.uk/art/artworks/blake-six-...  
4  http://www.tate.org.uk/art/artworks/blake-the-...

# 1) Viewing and Converting Types

In [ ]:
data.info()

In [11]:
data.dtypes

id                      int64
accession_number       object
artist                 object
artistRole             object
artistId                int64
title                  object
dateText               object
medium                 object
creditLine             object
year                  float64
acquisitionYear         int64
dimensions             object
width                   int64
height                  int64
depth                 float64
units                  object
inscription           float64
thumbnailCopyright    float64
thumbnailUrl           object
url                    object
dtype: object

In [12]:
data.acquisitionYear

0    1922
1    1922
2    1922
3    1922
4    1919
5    1919
6    1919
7    1919
8    1919
9    1919
Name: acquisitionYear, dtype: int64

In [13]:
data.acquisitionYear.astype(float)

0    1922.0
1    1922.0
2    1922.0
3    1922.0
4    1919.0
5    1919.0
6    1919.0
7    1919.0
8    1919.0
9    1919.0
Name: acquisitionYear, dtype: float64

In [14]:
data.acquisitionYear

0    1922
1    1922
2    1922
3    1922
4    1919
5    1919
6    1919
7    1919
8    1919
9    1919
Name: acquisitionYear, dtype: int64

In [15]:
data.acquisitionYear = data.acquisitionYear.astype(float)

In [16]:
data.dtypes

id                      int64
accession_number       object
artist                 object
artistRole             object
artistId                int64
title                  object
dateText               object
medium                 object
creditLine             object
year                  float64
acquisitionYear       float64
dimensions             object
width                   int64
height                  int64
depth                 float64
units                  object
inscription           float64
thumbnailCopyright    float64
thumbnailUrl           object
url                    object
dtype: object

In [17]:
fulldf = pd.read_csv("Data/artwork_data.csv", low_memory = False)

In [18]:
fulldf.head(2)

id accession_number         artist artistRole  artistId  \
0  1035           A00001  Blake, Robert     artist        38   
1  1036           A00002  Blake, Robert     artist        38   

                                               title        dateText  \
0  A Figure Bowing before a Seated Old Man with h...  date not known   
1  Two Drawings of Frightened Figures, Probably f...  date not known   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   

                            creditLine year  acquisitionYear  \
0  Presented by Mrs John Richmond 1922  NaN           1922.0   
1  Presented by Mrs John Richmond 1922  NaN           1922.0   

              dimensions width height  depth units inscription  \
0  support: 394 x 419 mm   394    419    NaN    mm         NaN   
1  support: 311 x 213 mm   311    213    NaN    mm         NaN   

  thumbnailCopyright                                       thumbnailUrl  \
0                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...

In [19]:
fulldf.dtypes

id                      int64
accession_number       object
artist                 object
artistRole             object
artistId                int64
title                  object
dateText               object
medium                 object
creditLine             object
year                   object
acquisitionYear       float64
dimensions             object
width                  object
height                 object
depth                 float64
units                  object
inscription            object
thumbnailCopyright     object
thumbnailUrl           object
url                    object
dtype: object

In [21]:
pd.to_numeric(fulldf.height)

ValueError: Unable to parse string "mm" at position 41339

### if we can't convert data type due to string, we can set `errors = coerce` parameter.

In [22]:
pd.to_numeric(fulldf.height, errors = "coerce")

0         419.0
1         213.0
2         467.0
3         394.0
4         335.0
          ...  
69196     305.0
69197     305.0
69198    2410.0
69199       NaN
69200     660.0
Name: height, Length: 69201, dtype: float64

In [23]:
# try to convert the row which had given issue
pd.to_numeric(fulldf.height, errors = "coerce")[41339]

nan

In [24]:
fulldf.height.dtype

dtype('O')

In [25]:
# convert the whole height to float
fulldf.height = pd.to_numeric(fulldf.height, errors = "coerce")

In [26]:
fulldf.height.dtype

dtype('float64')

-------

# 2) Aggregating Data

In [27]:
import pandas as pd

In [29]:
data = pd.read_csv("Data/artwork_sample.csv")

In [30]:
data.head(2)

id accession_number         artist artistRole  artistId  \
0  1035           A00001  Blake, Robert     artist        38   
1  1036           A00002  Blake, Robert     artist        38   

                                               title        dateText  \
0  A Figure Bowing before a Seated Old Man with h...  date not known   
1  Two Drawings of Frightened Figures, Probably f...  date not known   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   

                            creditLine  year  acquisitionYear  \
0  Presented by Mrs John Richmond 1922   NaN             1922   
1  Presented by Mrs John Richmond 1922   NaN             1922   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...

In [32]:
data.year

0       NaN
1       NaN
2    1785.0
3       NaN
4    1826.0
5    1826.0
6    1826.0
7    1826.0
8    1826.0
9    1826.0
Name: year, dtype: float64

In [35]:
# Aggregate functions on float data type column
data.year.min()

1785.0

In [36]:
data.year.max()

1826.0

In [37]:
data.year.sum()

12741.0

In [38]:
data.year.mean()

1820.142857142857

In [39]:
data.year.std()

15.496543393378316

In [42]:
# Aggregate functions on other data type too
# but mean() and std() won't work
data.artist.dtype

dtype('O')

In [44]:
data.artist.min()

'Blake, Robert'

In [45]:
data.artist.max()

'Blake, William'

In [47]:
data.artist.sum()

'Blake, RobertBlake, RobertBlake, RobertBlake, RobertBlake, WilliamBlake, WilliamBlake, WilliamBlake, WilliamBlake, WilliamBlake, William'

In [49]:
data.artist.std()

TypeError: could not convert string to float: 'Blake, Robert'

In [50]:
data.sum()

id                                                                10395
accession_number      A00001A00002A00003A00004A00005A00006A00007A000...
artist                Blake, RobertBlake, RobertBlake, RobertBlake, ...
artistRole            artistartistartistartistartistartistartistarti...
artistId                                                            386
title                 A Figure Bowing before a Seated Old Man with h...
dateText              date not knowndate not known?c.1785date not kn...
medium                Watercolour, ink, chalk and graphite on paper....
creditLine            Presented by Mrs John Richmond 1922Presented b...
year                                                              12741
acquisitionYear                                                   19202
dimensions            support: 394 x 419 mmsupport: 311 x 213 mmsupp...
width                                                              2821
height                                                          

## Using `agg()` method for aggregation

In [52]:
data.agg("min")

id                                                                 1035
accession_number                                                 A00001
artist                                                    Blake, Robert
artistRole                                                       artist
artistId                                                             38
title                 A Figure Bowing before a Seated Old Man with h...
dateText                                         1826–7, reprinted 1892
medium                                                Graphite on paper
creditLine                          Presented by Mrs John Richmond 1922
year                                                               1785
acquisitionYear                                                    1919
dimensions                                          image: 240 x 338 mm
width                                                               240
height                                                          

### Main advantages of using `agg()` is we can pass multiple agg type that we want to see as parameter
this will allow us to see at the glance across all columns, including string

In [53]:
data.agg(["min", "max"])

id accession_number          artist artistRole  artistId  \
min  1035           A00001   Blake, Robert     artist        38   
max  1044           A00010  Blake, William     artist        39   

                                                 title  \
min  A Figure Bowing before a Seated Old Man with h...   
max  Two Drawings of Frightened Figures, Probably f...   

                   dateText  \
min  1826–7, reprinted 1892   
max          date not known   

                                                medium  \
min                                  Graphite on paper   
max  Watercolour, ink, chalk and graphite on paper....   

                                            creditLine    year  \
min                Presented by Mrs John Richmond 1922  1785.0   
max  Purchased with the assistance of a special gra...  1826.0   

     acquisitionYear             dimensions  width  height  depth units  \
min             1919    image: 240 x 338 mm    240     213    NaN    mm   
max             1922  support: 394 x 419 mm    394     467    NaN    mm   

     inscription  thumbnailCopyright  \
min          NaN                 NaN   
max          NaN                 NaN   

                                          thumbnailUrl  \
min  http://www.tate.org.uk/art/images/work/A/A00/A...   
max  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                   url  
min  http://www.tate.org.uk/art/artworks/blake-a-fi...  
max  http://www.tate.org.uk/art/artworks/blake-two-...

In [54]:
data.agg(["min", "max", "mean", "std"])

id accession_number          artist artistRole   artistId  \
min   1035.00000           A00001   Blake, Robert     artist  38.000000   
max   1044.00000           A00010  Blake, William     artist  39.000000   
mean  1039.50000              NaN             NaN        NaN  38.600000   
std      3.02765              NaN             NaN        NaN   0.516398   

                                                  title  \
min   A Figure Bowing before a Seated Old Man with h...   
max   Two Drawings of Frightened Figures, Probably f...   
mean                                                NaN   
std                                                 NaN   

                    dateText  \
min   1826–7, reprinted 1892   
max           date not known   
mean                     NaN   
std                      NaN   

                                                 medium  \
min                                   Graphite on paper   
max   Watercolour, ink, chalk and graphite on paper....   
mean                                                NaN   
std                                                 NaN   

                                             creditLine         year  \
min                 Presented by Mrs John Richmond 1922  1785.000000   
max   Purchased with the assistance of a special gra...  1826.000000   
mean                                                NaN  1820.142857   
std                                                 NaN    15.496543   

      acquisitionYear             dimensions       width      height  depth  \
min       1919.000000    image: 240 x 338 mm  240.000000  213.000000    NaN   
max       1922.000000  support: 394 x 419 mm  394.000000  467.000000    NaN   
mean      1920.200000                    NaN  282.100000  351.500000    NaN   
std          1.549193                    NaN   55.562677   66.818577    NaN   

     units  inscription  thumbnailCopyright  \
min     mm          NaN                 NaN   
max     mm          NaN                 NaN   
mean   NaN          NaN                 NaN   
std    NaN          NaN                 NaN   

                                           thumbnailUrl  \
min   http://www.tate.org.uk/art/images/work/A/A00/A...   
max   http://www.tate.org.uk/art/images/work/A/A00/A...   
mean                                                NaN   
std                                                 NaN   

                                                    url  
min   http://www.tate.org.uk/art/artworks/blake-a-fi...  
max   http://www.tate.org.uk/art/artworks/blake-two-...  
mean                                                NaN  
std                                                 NaN

In [55]:
data.describe() #describe didn't work for strings

id   artistId         year  acquisitionYear       width  \
count    10.00000  10.000000     7.000000        10.000000   10.000000   
mean   1039.50000  38.600000  1820.142857      1920.200000  282.100000   
std       3.02765   0.516398    15.496543         1.549193   55.562677   
min    1035.00000  38.000000  1785.000000      1919.000000  240.000000   
25%    1037.25000  38.000000  1826.000000      1919.000000  242.250000   
50%    1039.50000  39.000000  1826.000000      1919.000000  244.500000   
75%    1041.75000  39.000000  1826.000000      1922.000000  316.250000   
max    1044.00000  39.000000  1826.000000      1922.000000  394.000000   

           height  depth  inscription  thumbnailCopyright  
count   10.000000    0.0          0.0                 0.0  
mean   351.500000    NaN          NaN                 NaN  
std     66.818577    NaN          NaN                 NaN  
min    213.000000    NaN          NaN                 NaN  
25%    335.000000    NaN          NaN                 NaN  
50%    339.000000    NaN          NaN                 NaN  
75%    380.500000    NaN          NaN                 NaN  
max    467.000000    NaN          NaN                 NaN

### Aggregation by rows

In [58]:
data.agg("mean", axis = "columns")

Series([], dtype: float64)

# 3) Normalizing Data

- Normalizing means we want to adjust the values in the columns and change their scale
- Normalizing is usually done on `Numeric` data type.

In [61]:
data.dtypes

id                      int64
accession_number       object
artist                 object
artistRole             object
artistId                int64
title                  object
dateText               object
medium                 object
creditLine             object
year                  float64
acquisitionYear         int64
dimensions             object
width                   int64
height                  int64
depth                 float64
units                  object
inscription           float64
thumbnailCopyright    float64
thumbnailUrl           object
url                    object
dtype: object

In [62]:
data.height.mean()

351.5

In [63]:
data.height

0    419
1    213
2    467
3    394
4    335
5    338
6    334
7    340
8    335
9    340
Name: height, dtype: int64

In [64]:
data.height.min()

213

In [65]:
data.height.max()

467

In [66]:
data.height.mean()

351.5

In [67]:
data.height.std()

66.81857692455162

## Standardization
- x - mean / std

In [68]:
height = data.height

In [71]:
# step 1
norm = (height - height.mean()) / height.std()

In [72]:
norm

0    1.010198
1   -2.072777
2    1.728561
3    0.636051
4   -0.246937
5   -0.202040
6   -0.261903
7   -0.172108
8   -0.246937
9   -0.172108
Name: height, dtype: float64

In [73]:
# step 2: if we also want to normalize the value into, between 0 and 1
minmax = (height - height.min()) / (height.max() - height.min())
minmax

0    0.811024
1    0.000000
2    1.000000
3    0.712598
4    0.480315
5    0.492126
6    0.476378
7    0.500000
8    0.480315
9    0.500000
Name: height, dtype: float64

In [74]:
minmax.min()

0.0

In [75]:
minmax.max()

1.0

In [78]:
data.height = minmax

In [79]:
# create as new column
data["standardized_height"] = norm

In [80]:
data.standardized_height

0    1.010198
1   -2.072777
2    1.728561
3    0.636051
4   -0.246937
5   -0.202040
6   -0.261903
7   -0.172108
8   -0.246937
9   -0.172108
Name: standardized_height, dtype: float64

In [81]:
data.head(2)

id accession_number         artist artistRole  artistId  \
0  1035           A00001  Blake, Robert     artist        38   
1  1036           A00002  Blake, Robert     artist        38   

                                               title        dateText  \
0  A Figure Bowing before a Seated Old Man with h...  date not known   
1  Two Drawings of Frightened Figures, Probably f...  date not known   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   

                            creditLine  year  ...             dimensions  \
0  Presented by Mrs John Richmond 1922   NaN  ...  support: 394 x 419 mm   
1  Presented by Mrs John Richmond 1922   NaN  ...  support: 311 x 213 mm   

  width    height  depth  units inscription  thumbnailCopyright  \
0   394  0.811024    NaN     mm         NaN                 NaN   
1   311  0.000000    NaN     mm         NaN                 NaN   

                                        thumbnailUrl  \
0  http://www.tate.org.uk/art/images/work/A/A00/A...   
1  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url standardized_height  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...            1.010198  
1  http://www.tate.org.uk/art/artworks/blake-two-...           -2.072777  

[2 rows x 21 columns]

---------

# 4) Transforming Data

In [87]:
data = pd.read_csv("Data/artwork_sample.csv")
data.head(2)

id accession_number         artist artistRole  artistId  \
0  1035           A00001  Blake, Robert     artist        38   
1  1036           A00002  Blake, Robert     artist        38   

                                               title        dateText  \
0  A Figure Bowing before a Seated Old Man with h...  date not known   
1  Two Drawings of Frightened Figures, Probably f...  date not known   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   

                            creditLine  year  acquisitionYear  \
0  Presented by Mrs John Richmond 1922   NaN             1922   
1  Presented by Mrs John Richmond 1922   NaN             1922   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...

### Transform column using lambda

In [92]:
#just print out values
data.height.transform(lambda x: print(x))

419
213
467
394
335
338
334
340
335
340


0    None
1    None
2    None
3    None
4    None
5    None
6    None
7    None
8    None
9    None
Name: height, dtype: object

In [88]:
#transforms the value by dividing 10 => cm to meters
data.height.transform(lambda x : x /10)

0    41.9
1    21.3
2    46.7
3    39.4
4    33.5
5    33.8
6    33.4
7    34.0
8    33.5
9    34.0
Name: height, dtype: float64

In [89]:
data.head(2)

id accession_number         artist artistRole  artistId  \
0  1035           A00001  Blake, Robert     artist        38   
1  1036           A00002  Blake, Robert     artist        38   

                                               title        dateText  \
0  A Figure Bowing before a Seated Old Man with h...  date not known   
1  Two Drawings of Frightened Figures, Probably f...  date not known   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   

                            creditLine  year  acquisitionYear  \
0  Presented by Mrs John Richmond 1922   NaN             1922   
1  Presented by Mrs John Richmond 1922   NaN             1922   

              dimensions  width  height  depth units  inscription  \
0  support: 394 x 419 mm    394     419    NaN    mm          NaN   
1  support: 311 x 213 mm    311     213    NaN    mm          NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...

### Tranform got powerful when lambda is used with groupby

In [93]:
data.groupby("artist")

In [94]:
data.groupby("artist").transform("nunique")

id  accession_number  artistRole  artistId  title  dateText  medium  \
0   4                 4           1         1      4         2       3   
1   4                 4           1         1      4         2       3   
2   4                 4           1         1      4         2       3   
3   4                 4           1         1      4         2       3   
4   6                 6           1         1      6         1       1   
5   6                 6           1         1      6         1       1   
6   6                 6           1         1      6         1       1   
7   6                 6           1         1      6         1       1   
8   6                 6           1         1      6         1       1   
9   6                 6           1         1      6         1       1   

   creditLine  year  acquisitionYear  dimensions  width  height  depth  units  \
0           1     1                1           4      4       4      0      1   
1           1     1                1           4      4       4      0      1   
2           1     1                1           4      4       4      0      1   
3           1     1                1           4      4       4      0      1   
4           1     1                1           6      5       4      0      1   
5           1     1                1           6      5       4      0      1   
6           1     1                1           6      5       4      0      1   
7           1     1                1           6      5       4      0      1   
8           1     1                1           6      5       4      0      1   
9           1     1                1           6      5       4      0      1   

   inscription  thumbnailCopyright  thumbnailUrl  url  
0            0                   0             4    4  
1            0                   0             4    4  
2            0                   0             4    4  
3            0                   0             4    4  
4            0                   0             6    6  
5            0                   0             6    6  
6            0                   0             6    6  
7            0                   0             6    6  
8            0                   0             6    6  
9            0                   0             6    6

In [98]:
data.groupby("artist")["height"]

In [99]:
data.groupby("artist")["height"].transform("mean")

0    373.25
1    373.25
2    373.25
3    373.25
4    337.00
5    337.00
6    337.00
7    337.00
8    337.00
9    337.00
Name: height, dtype: float64

we can see that artist "Blake, Robert" artwork peices have mean of 373.25 

In [109]:
data.artist

0     Blake, Robert
1     Blake, Robert
2     Blake, Robert
3     Blake, Robert
4    Blake, William
5    Blake, William
6    Blake, William
7    Blake, William
8    Blake, William
9    Blake, William
Name: artist, dtype: object

In [110]:
data.groupby("artist")["height"].mean()

artist
Blake, Robert     373.25
Blake, William    337.00
Name: height, dtype: float64

In [113]:
# creating transformed data into a new column
data["mean_height_by_artist"] = data.groupby("artist")["height"].transform("mean")

In [114]:
data.head()

id accession_number          artist artistRole  artistId  \
0  1035           A00001   Blake, Robert     artist        38   
1  1036           A00002   Blake, Robert     artist        38   
2  1037           A00003   Blake, Robert     artist        38   
3  1038           A00004   Blake, Robert     artist        38   
4  1039           A00005  Blake, William     artist        39   

                                               title                dateText  \
0  A Figure Bowing before a Seated Old Man with h...          date not known   
1  Two Drawings of Frightened Figures, Probably f...          date not known   
2  The Preaching of Warning. Verso: An Old Man En...                 ?c.1785   
3     Six Drawings of Figures with Outstretched Arms          date not known   
4  The Circle of the Lustful: Francesca da Rimini...  1826–7, reprinted 1892   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   
2        Graphite on paper. Verso: graphite on paper   
3                                  Graphite on paper   
4                            Line engraving on paper   

                                          creditLine    year  ...  \
0                Presented by Mrs John Richmond 1922     NaN  ...   
1                Presented by Mrs John Richmond 1922     NaN  ...   
2                Presented by Mrs John Richmond 1922  1785.0  ...   
3                Presented by Mrs John Richmond 1922     NaN  ...   
4  Purchased with the assistance of a special gra...  1826.0  ...   

              dimensions width  height  depth  units inscription  \
0  support: 394 x 419 mm   394     419    NaN     mm         NaN   
1  support: 311 x 213 mm   311     213    NaN     mm         NaN   
2  support: 343 x 467 mm   343     467    NaN     mm         NaN   
3  support: 318 x 394 mm   318     394    NaN     mm         NaN   
4    image: 243 x 335 mm   243     335    NaN     mm         NaN   

   thumbnailCopyright                                       thumbnailUrl  \
0                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
2                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
3                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
4                 NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url mean_height_by_artist  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...                373.25  
1  http://www.tate.org.uk/art/artworks/blake-two-...                373.25  
2  http://www.tate.org.uk/art/artworks/blake-the-...                373.25  
3  http://www.tate.org.uk/art/artworks/blake-six-...                373.25  
4  http://www.tate.org.uk/art/artworks/blake-the-...                337.00  

[5 rows x 21 columns]

-------------------


# 5) Filtering Data

In [116]:
data = pd.read_csv("Data/artwork_data.csv", low_memory = False)

In [117]:
data.head(2)

id accession_number         artist artistRole  artistId  \
0  1035           A00001  Blake, Robert     artist        38   
1  1036           A00002  Blake, Robert     artist        38   

                                               title        dateText  \
0  A Figure Bowing before a Seated Old Man with h...  date not known   
1  Two Drawings of Frightened Figures, Probably f...  date not known   

                                              medium  \
0  Watercolour, ink, chalk and graphite on paper....   
1                                  Graphite on paper   

                            creditLine year  acquisitionYear  \
0  Presented by Mrs John Richmond 1922  NaN           1922.0   
1  Presented by Mrs John Richmond 1922  NaN           1922.0   

              dimensions width height  depth units inscription  \
0  support: 394 x 419 mm   394    419    NaN    mm         NaN   
1  support: 311 x 213 mm   311    213    NaN    mm         NaN   

  thumbnailCopyright                                       thumbnailUrl  \
0                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   
1                NaN  http://www.tate.org.uk/art/images/work/A/A00/A...   

                                                 url  
0  http://www.tate.org.uk/art/artworks/blake-a-fi...  
1  http://www.tate.org.uk/art/artworks/blake-two-...

### Filtering by columns

In [119]:
data.filter(items=["id", "artist"])

id              artist
0        1035       Blake, Robert
1        1036       Blake, Robert
2        1037       Blake, Robert
3        1038       Blake, Robert
4        1039      Blake, William
...       ...                 ...
69196  122960  P-Orridge, Genesis
69197  122961  P-Orridge, Genesis
69198  121181        Hatoum, Mona
69199  112306       Creed, Martin
69200  127035   Brunias, Agostino

[69201 rows x 2 columns]

### Filtering by `like` parameters
- `like` parameter is case sensitive

In [121]:
data.filter(like = 'year')
# in here we can only get "year" column, not "acquisitionYear" column

year
0       NaN
1       NaN
2      1785
3       NaN
4      1826
...     ...
69196  1975
69197  1976
69198  1996
69199  2000
69200  1764

[69201 rows x 1 columns]

In [122]:
data.filter(like="artist")

artist artistRole  artistId
0           Blake, Robert     artist        38
1           Blake, Robert     artist        38
2           Blake, Robert     artist        38
3           Blake, Robert     artist        38
4          Blake, William     artist        39
...                   ...        ...       ...
69196  P-Orridge, Genesis     artist     16646
69197  P-Orridge, Genesis     artist     16646
69198        Hatoum, Mona     artist      2365
69199       Creed, Martin     artist      2760
69200   Brunias, Agostino     artist      3824

[69201 rows x 3 columns]

### using `Regex` for full control

In [124]:
#now we can get case insensitive year filter
data.filter(regex = "(?i)year")

year  acquisitionYear
0       NaN           1922.0
1       NaN           1922.0
2      1785           1922.0
3       NaN           1922.0
4      1826           1919.0
...     ...              ...
69196  1975           2013.0
69197  1976           2013.0
69198  1996           2013.0
69199  2000           2013.0
69200  1764           2013.0

[69201 rows x 2 columns]

In [125]:
# getting rows
data.filter(axis = 0, regex = "^100.$")

id accession_number                          artist artistRole  \
1000  14704           A01004  Turner, Joseph Mallord William     artist   
1001  14705           A01005  Turner, Joseph Mallord William     artist   
1002  14706           A01006  Turner, Joseph Mallord William     artist   
1003  14707           A01007  Turner, Joseph Mallord William     artist   
1004  14708           A01008  Turner, Joseph Mallord William     artist   
1005  14709           A01009  Turner, Joseph Mallord William     artist   
1006  14710           A01010  Turner, Joseph Mallord William     artist   
1007  14711           A01011  Turner, Joseph Mallord William     artist   
1008  14712           A01012  Turner, Joseph Mallord William     artist   
1009  14713           A01013  Turner, Joseph Mallord William     artist   

      artistId                                              title dateText  \
1000       558                               Hedging and Ditching     1812   
1001       558     Hedging and Ditching, engraved by J.C. Easling     1812   
1002       558                                          River Wye     1812   
1003       558                    River Wye, engraved by W. Annis     1812   
1004       558            Chain of Alps from Grenoble to Chamberi     1812   
1005       558  Chain of Alps from Grenoble to Chamberi, engra...     1812   
1006       558                                       Mer de Glace     1812   
1007       558                                       Mer de Glace     1812   
1008       558                            Rivaux Abbey, Yorkshire     1812   
1009       558               Rivaux Abbey, engraved by Henry Dawe     1812   

                              medium  \
1000  Etching and engraving on paper   
1001  Etching and mezzotint on paper   
1002                  Print on paper   
1003  Etching and mezzotint on paper   
1004  Etching and engraving on paper   
1005  Etching and mezzotint on paper   
1006  Etching and engraving on paper   
1007  Etching and mezzotint on paper   
1008  Etching and engraving on paper   
1009  Etching and mezzotint on paper   

                                             creditLine  year  \
1000  Presented by A. Acland Allen through the Art F...  1812   
1001  Presented by A. Acland Allen through the Art F...  1812   
1002  Presented by A. Acland Allen through the Art F...  1812   
1003  Presented by A. Acland Allen through the Art F...  1812   
1004  Presented by A. Acland Allen through the Art F...  1812   
1005  Presented by A. Acland Allen through the Art F...  1812   
1006  Presented by A. Acland Allen through the Art F...  1812   
1007  Presented by A. Acland Allen through the Art F...  1812   
1008  Presented by A. Acland Allen through the Art F...  1812   
1009  Presented by A. Acland Allen through the Art F...  1812   

      acquisitionYear           dimensions width height  depth units  \
1000           1925.0  image: 184 x 259 mm   184    259    NaN    mm   
1001           1925.0  image: 186 x 262 mm   186    262    NaN    mm   
1002           1925.0  image: 183 x 264 mm   183    264    NaN    mm   
1003           1925.0  image: 184 x 265 mm   184    265    NaN    mm   
1004           1925.0  image: 178 x 259 mm   178    259    NaN    mm   
1005           1925.0  image: 178 x 261 mm   178    261    NaN    mm   
1006           1925.0  image: 178 x 253 mm   178    253    NaN    mm   
1007           1925.0  image: 179 x 257 mm   179    257    NaN    mm   
1008           1925.0  image: 182 x 265 mm   182    265    NaN    mm   
1009           1925.0  image: 183 x 266 mm   183    266    NaN    mm   

     inscription thumbnailCopyright  \
1000         NaN                NaN   
1001         NaN                NaN   
1002         NaN                NaN   
1003         NaN                NaN   
1004         NaN                NaN   
1005         NaN                NaN   
1006         NaN                NaN   
1007         NaN                NaN   
1008         NaN       

In [128]:
# complex regex for even values
data.filter(regex = "^100(0|2|4|6|8)$", axis = 0)

id accession_number                          artist artistRole  \
1000  14704           A01004  Turner, Joseph Mallord William     artist   
1002  14706           A01006  Turner, Joseph Mallord William     artist   
1004  14708           A01008  Turner, Joseph Mallord William     artist   
1006  14710           A01010  Turner, Joseph Mallord William     artist   
1008  14712           A01012  Turner, Joseph Mallord William     artist   

      artistId                                    title dateText  \
1000       558                     Hedging and Ditching     1812   
1002       558                                River Wye     1812   
1004       558  Chain of Alps from Grenoble to Chamberi     1812   
1006       558                             Mer de Glace     1812   
1008       558                  Rivaux Abbey, Yorkshire     1812   

                              medium  \
1000  Etching and engraving on paper   
1002                  Print on paper   
1004  Etching and engraving on paper   
1006  Etching and engraving on paper   
1008  Etching and engraving on paper   

                                             creditLine  year  \
1000  Presented by A. Acland Allen through the Art F...  1812   
1002  Presented by A. Acland Allen through the Art F...  1812   
1004  Presented by A. Acland Allen through the Art F...  1812   
1006  Presented by A. Acland Allen through the Art F...  1812   
1008  Presented by A. Acland Allen through the Art F...  1812   

      acquisitionYear           dimensions width height  depth units  \
1000           1925.0  image: 184 x 259 mm   184    259    NaN    mm   
1002           1925.0  image: 183 x 264 mm   183    264    NaN    mm   
1004           1925.0  image: 178 x 259 mm   178    259    NaN    mm   
1006           1925.0  image: 178 x 253 mm   178    253    NaN    mm   
1008           1925.0  image: 182 x 265 mm   182    265    NaN    mm   

     inscription thumbnailCopyright  \
1000         NaN                NaN   
1002         NaN                NaN   
1004         NaN                NaN   
1006         NaN                NaN   
1008         NaN                NaN   

                                           thumbnailUrl  \
1000  http://www.tate.org.uk/art/images/work/A/A01/A...   
1002  http://www.tate.org.uk/art/images/work/A/A01/A...   
1004  http://www.tate.org.uk/art/images/work/A/A01/A...   
1006  http://www.tate.org.uk/art/images/work/A/A01/A...   
1008  http://www.tate.org.uk/art/images/work/A/A01/A...   

                                                    url  
1000  http://www.tate.org.uk/art/artworks/turner-hed...  
1002  http://www.tate.org.uk/art/artworks/turner-riv...  
1004  http://www.tate.org.uk/art/artworks/turner-cha...  
1006  http://www.tate.org.uk/art/artworks/turner-mer...  
1008  http://www.tate.org.uk/art/artworks/turner-riv...

------------